sycl/test/fpga_tests/arbitrary_template_arg.cpp

```diff
@@ -0,0 +1,38 @@
+// RUN: %clangxx -fsycl -fsyntax-only -Xclang -verify %s -Xclang -verify-ignore-unexpected=note,warning
+// expected-no-diagnostics
+
+// This test performs basic check of SYCL FPGA arbitrary template argument list.
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_utils.hpp>
+#include <iostream>
+
+enum class enumClass { first, second };
+
+struct testIntID;
+template <int _N> struct testInt : std::integral_constant<int, _N> {
+  using type_id = testIntID;
+};
+
+struct testEnumID;
+template <enumClass _N>
+struct testEnum : std::integral_constant<enumClass, _N> {
+  using type_id = testEnumID;
+};
+
+template <int ExpectedIntValue, enumClass ExpectedEnumValue, class... _Params>
+void func() {
+  static_assert(sycl::INTEL::_GetValue<testInt<0>, _Params...>::value ==
+                ExpectedIntValue);
+  static_assert(
+      sycl::INTEL::_GetValue<testEnum<enumClass::first>, _Params...>::value ==
+      ExpectedEnumValue);
+}
+
+int main() {
+  func<0, enumClass::first>();
+  func<1, enumClass::first, testInt<1>>();
+  func<0, enumClass::second, testEnum<enumClass::second>>();
+  func<1, enumClass::second, testInt<1>, testEnum<enumClass::second>>();
+  func<1, enumClass::second, testEnum<enumClass::second>, testInt<1>>();
```

Comment:
`burst_coalesce_impl`, `cache`, `prefetch_impl` and `statically_coalesce_impl` should be covered by tests too.

Reply:
This new test is created only for fpga_utils.hpp.
I believe functionality of fpga_lsu.hpp is tested by `[2021-07-07T18:51:41.321Z] PASS: SYCL :: Basic/fpga_tests/fpga_lsu.cpp (59 of 530)` in the compiled test (Jenkins/Precommit: http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FLin%2FLLVM_Test_Suite/detail/LLVM_Test_Suite/5893/pipeline).

Reply:
OK, sounds good to me.

Reply:
Also, I just found llvm/sycl/test/extensions/fpga.cpp tests all FPGA interfaces including fpgs_lsu. See line 13-25 where `burst_coalesce_impl`, `cache`, `prefetch_impl`, and `statically_coalesce_impl` are tested.